crates/fluvio-spu/src/kv/consumer.rs

```diff
@@ -0,0 +1,295 @@
+    LeaderKVStorage, FollowerNotifier, LeaderReplicaState, LeaderReplicaLog,
+};
+
+pub(crate) type TimestampSecs = u64;
+
+const DEFAULT_FLUSH_THRESHOLD: usize = 100;
+
+#[derive(Debug, Default)]
+pub(crate) struct SharedConsumerStorages(Arc<RwLock<HashMap<ReplicaKey, SharableConsumerStorage>>>);
+
+#[derive(Debug, Clone)]
+pub(crate) struct SharableConsumerStorage(Arc<RwLock<ConsumerStorage>>);
+
+#[derive(Debug, Hash, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Encoder, Decoder)]
+pub(crate) struct Key {
+    replica_id: ReplicaKey,
+    consumer_id: String,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encoder, Decoder)]
+pub(crate) struct Consumer {
+    pub offset: i64,
+    pub ttl: Duration,
+    pub expire_time: TimestampSecs,
+}
+
+#[derive(Debug)]
+pub(crate) struct ConsumerStorage {
+    kv: LeaderKVStorage<Key, Consumer, FileReplica>,
+    flush_threshold: usize,
+    changes_since_flush: usize,
+}
+
+impl SharedConsumerStorages {
+    pub(crate) async fn get_or_insert(
+        &self,
+        replica: &LeaderReplicaState<FileReplica>,
+        notifier: &Arc<FollowerNotifier>,
+    ) -> Result<SharableConsumerStorage> {
+        let mut write = self.0.write().await;
+        match write.entry(replica.id().clone()) {
+            Entry::Occupied(entry) => Ok(entry.get().clone()),
+            Entry::Vacant(entry) => {
+                let mut storage = ConsumerStorage::new(replica.clone(), notifier.clone());
+                storage.kv.sync_from_log().await?;
+                let shared: SharableConsumerStorage = storage.into();
+                entry.insert(shared.clone());
+                Ok(shared)
+            }
+        }
+    }
+}
+
+impl ConsumerStorage {
+    pub fn new(
+        replica: LeaderReplicaState<FileReplica>,
+        follower_notifier: Arc<FollowerNotifier>,
+    ) -> Self {
+        Self::with(replica, follower_notifier, DEFAULT_FLUSH_THRESHOLD)
+    }
+
+    pub fn with(
+        replica: LeaderReplicaState<FileReplica>,
+        follower_notifier: Arc<FollowerNotifier>,
+        flush_threshold: usize,
+    ) -> Self {
+        Self {
+            kv: LeaderKVStorage::new(LeaderReplicaLog::new(replica, follower_notifier)),
+            flush_threshold,
+            changes_since_flush: Default::default(),
+        }
+    }
+
+    async fn maybe_flush(&mut self) -> Result<()> {
+        if self.changes_since_flush > self.flush_threshold {
+            self.kv.flush().await?;
+            self.changes_since_flush = Default::default();
+        }
+        Ok(())
+    }
+}
+
+impl KVStorage<Key, Consumer> for ConsumerStorage {
+    async fn get(&self, key: &Key) -> Result<Option<Consumer>> {
+        trace!(?key, "get");
+        self.kv.get(key).await.map(|op| op.filter(is_not_expired))
+    }
+
+    async fn delete(&mut self, key: &Key) -> Result<()> {
+        trace!(?key, "delete");
+        let result = self.kv.delete(key).await;
+        self.changes_since_flush.add_assign(1);
```

Comment:
why need to have flush logic? Shouldn't that be built into KV?

Reply:
Flush logic is on KV, but since it is not a cheap operation (memory and latency), the control over when it is used is put on a higher level to give control to the business logic level. For example, for consumers offset we might want to flush with one frequency, but for other entities with different. That's why the inner KV storage does not do it internally.

Reply:
fair enough